mlir/lib/Dialect/LLVMIR/IR/GENXDialect.cpp

```diff
@@ -60,6 +63,153 @@ LogicalResult GENXDialect::verifyOperationAttribute(Operation *op,
+      return Type();
+    }
+  } else if (auto t = type.dyn_cast<IntegerType>()) {
+    if (!llvm::is_contained({1u, 8u, 16u, 32u, 64u}, t.getWidth())) {
+      parser.emitError(typeLoc,
+                       "only 1/8/16/32/64-bit integer type allowed but found ")
+          << type;
+      return Type();
+    }
+  } else if (auto t = type.dyn_cast<VectorType>()) {
+    if (t.getRank() != 1) {
+      parser.emitError(typeLoc, "only 1-D vector allowed but found ") << t;
+      return Type();
+    }
+    if (t.getNumElements() > 4) {
+      parser.emitError(
+          typeLoc, "vector length has to be less than or equal to 4 but found ")
+          << t.getNumElements();
+      return Type();
+    }
+  } else {
+    parser.emitError(typeLoc, "cannot use ")
+        << type << " to compose GENX types";
+    return Type();
+  }
+
+  return type;
+}
+
+/// Parses the next keyword in `parser` as an enumerator.
+template <typename EnumClass, typename ParserType>
+static ParseResult parseEnumKeywordAttr(EnumClass &value, ParserType &parser,
+                                        StringRef attrName) {
+  StringRef keyword;
+  SmallVector<NamedAttribute, 1> attr;
+  auto loc = parser.getCurrentLocation();
+  if (parser.parseKeyword(&keyword))
+    return failure();
+  if (std::optional<EnumClass> attr = GENX::symbolizeEnum<EnumClass>(keyword)) {
+    value = *attr;
+    return success();
+  }
+  return parser.emitError(loc, "invalid ")
+         << attrName << " attribute specification: " << keyword;
+}
+
+// joint-matrix-type ::= `!genx.jointmatrix` `<`rows `x` cols `x` element-type
+//                                           `,` layout `,` scope`>`
+static Type parseJointMatrixType(GENXDialect const &dialect,
+                                 DialectAsmParser &parser) {
+  if (parser.parseLess())
+    return Type();
+
+  SmallVector<int64_t, 2> dims;
+  SMLoc countLoc = parser.getCurrentLocation();
+  if (parser.parseDimensionList(dims, /*allowDynamic=*/false))
+    return Type();
+
+  if (dims.size() != 2) {
+    parser.emitError(countLoc, "expected rows and columns size");
+    return Type();
+  }
+
+  auto elementTy = parseAndVerifyType(dialect, parser);
+  if (!elementTy)
+    return Type();
+  MatrixLayout matrixLayout;
+  if (parser.parseComma() ||
+      parseEnumKeywordAttr(matrixLayout, parser, "matrixLayout <id>"))
+    return Type();
+  Scope scope;
+  if (parser.parseComma() || parseEnumKeywordAttr(scope, parser, "scope <id>"))
+    return Type();
+  if (parser.parseGreater())
+    return Type();
+  return JointMatrixType::get(elementTy, scope, dims[0], dims[1], matrixLayout);
+}
+
+// Specialize this function to parse each of the parameters that define an
+// ImageType. By default it assumes this is an enum type.
+template <typename ValTy>
+static std::optional<ValTy> parseAndVerify(GENXDialect const &dialect,
+                                           DialectAsmParser &parser) {
```

Comment:
Do we need this function? It seems not called from this file, and from the comment, seems to be for ImageType, which is not defined in GENX.

Reply:
removed